### PR TITLE
Add a E2E based performance test case

### PR DESCRIPTION
**Pull request by CareF**
## Description

This PR reimplement `cull_opacity_perf__timeline_summary` test as `cull_opacity_perf__e2e_summary` using E2E, as a trial for running performance test using E2E that's host-independent in terms of driving the test. 

## Related Issues

This is the first migration of performance test for go/flutter-nonhost
This is also a first step to flutter/flutter#38329 for the new test case is self-driven. 

## Tests

The PR itself is adding a test case 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/integration_tests/release_smoke_test/pubspec.yaml
Comment (CareF) [inline dev/integration_tests/release_smoke_test/pubspec.yaml:21]: This change is temporary for demo purposes only. After flutter/plugins#2873 is landed this will be updaged.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (liyuqian) [inline dev/devicelab/lib/tasks/perf_tests.dart:398]: I'm not completely sure why this `internalRun` needs to be rewritten... what lines of code should I focus on reading?

Comment (CareF) [reply dev/devicelab/lib/tasks/perf_tests.dart:398]: L439 ~ 446, the key list. This E2E version cannot get vsync related metrics.

Comment (CareF) [reply dev/devicelab/lib/tasks/perf_tests.dart:398]: Other than that, I'm expecting E2E to share `testDriver` and to have different timeline file name. It doesn't make sense to still name the file `**.timeline_summary.json` since its of different format.

Comment (CareF) [reply dev/devicelab/lib/tasks/perf_tests.dart:398]: Mentioning this, I think '--trace-startup' is no longer needed. I'll delay the update to keep the above line numbers valid.

**Comment by CareF** [comment]
A few things to address: 
- part of `dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart` and `dev/benchmarks/macrobenchmarks/test/util.dart` may be better go to E2E; 
- Should we or when should we replace the original perf test case: For this one there will be duplicate on the flutter-build webpage, but I don't think eventually we need both `flutter_driver` version and E2E version running.

File: dev/devicelab/manifest.yaml
Comment (dnfield) [inline dev/devicelab/manifest.yaml:169]: nit: prefer "linux/android" when possible.

Comment (CareF) [reply dev/devicelab/manifest.yaml]: Done

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (liyuqian) [reply dev/devicelab/lib/tasks/perf_tests.dart:398]: To minimize duplicate code, maybe we can let `internalRun` take options like `bool reportsVsyncMetrics = true, bool needsTraceStartup = true`, and then just use `internalRun(reportsVsyncMetrics: false, needsTraceStartup: false)` in the e2e test?

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:108]: Nit: `FrameTimingSummarizer` seems to be a more informative name.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:1]: This file probably also belongs to `test_driver` as this is unlikely to be used as a unit test. BTW, I guess we'll update `WidgetTester` to `WidgetController` in this file?

File: dev/benchmarks/macrobenchmarks/test/cull_opacity_perf_e2e.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/cull_opacity_perf_e2e.dart:12]: This file probably belongs to `test_driver` as this is unlikely to be used as a unit test.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:20]: When this PR is out of draft, we should try to minimize duplicate code as it's hard to maintain.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:1]: I'm not expecting to use that demo very soon but if you think I should, maybe I'd better change this PR back to draft as that demo is far from use-able.

Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:1]: And the reason I'm putting it in `test/` rather than test_driver is that E2E recommends so (https://pub.dev/packages/e2e#test-locations), and the test file (`cull_opacity_perf_e2e.dart`) should be able to run using `flutter test`.

File: dev/benchmarks/macrobenchmarks/test/cull_opacity_perf_e2e.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/cull_opacity_perf_e2e.dart:12]: Same reason above.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:20]: If you mean re-use the warning string from `packages/flutter_driver/lib/src/driver/vmservice_driver.dart`, that one is private and need an extra line break to align the warning message.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:20]: And although it's out of draft, I'm still thinking it as experimental and expecting to see how it will go with device lab. If it's going on well, I think it will worth discussing if `FrameTimingSummarizer` and `watchPerformance` should go to e2e or `flutter_test`.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:1]: Changing from `WidgetTester` to `WidgetController` for `driverOps` and `setupOps` for now, but this specific case it's not used.

**Comment by CareF** [comment]
I may also need to update the macrobenchmark/readme.md, which is also the case for #61388. That will be done on whichever of the two PR who lands later.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [reply dev/benchmarks/macrobenchmarks/test/util.dart:20]: As [`FrameTiming`] is in the SDK, I wonder if we can also put `FrameTimingSummarizer ` in the SDK just in case that developers also want to monitor their app performance in the wild (e.g., https://github.com/flutter/flutter/issues/37437). What do you think, @Hixie ?

[`FrameTiming`]: https://api.flutter.dev/flutter/dart-ui/FrameTiming-class.html

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:180]: Nit: I feel it's safer to use a simple O(n) scanning algorithm here than a cleverer O(log n) binary search. We don't have unit test coverage for this function so it might be easy for someone to break this in the future without noticing it.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (liyuqian) [reply dev/devicelab/lib/tasks/perf_tests.dart:458]: @CareF : how do you think about the last comment about reducing duplicate code?

File: dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart:6]: Is it true that we don't need this file if flutter/plugins#2859 landed? Since this PR doesn't need Timeline, I wonder if we only need a part of flutter/plugins#2859 without the Timeline support.

**Review by liyuqian — COMMENTED** [review]
I think this is good to go as soon as we minimize the duplicate code to make it a little easier to maintain in the future. Let's try to have this running in our device lab soon :)

File: dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart:6]: My bad. This comment comes here for my first trial version. Now this file is a little different with flutter/plugins #2859 but the larger idea is similar.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:20]: `flutter_driver` cannot import things from `macrobenchmarks` because the latter is not a published package and is not a dependency of `flutter_driver`.  There's another similar thing in https://github.com/flutter/flutter/blob/07caa0fbfe5186cc8a11902cc46377cebd592088/packages/flutter_test/lib/src/widget_tester.dart#L324-L334 . I think it might be better to put in `flutter_test`, if we have to.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:40]: I interpret `driver` here as the operation to *drive* the test, rather than from `flutter_driver`. What about call them `driverCallback` and `setupCallback`. And maybe i should also `typedef` the callback.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:125]: Then `kBuildBudget` would have to be a getter (or a const) so user won't expect to change it to change the related result.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done with the change, but let me know if `kBuildBudget` should still be a value that can be changed by customer, in which case it will need to be a member variable of the class.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:224]: Done. A little sad when I finally find small algos are useful somewhere but it turns out not a good practice :(

File: dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test_driver/e2e_test.dart]: Removed the comment. See new comment in flutter/plugins#2859

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (CareF) [reply dev/devicelab/lib/tasks/perf_tests.dart]: I had a slightly different design but still reduce the duplicate. Let me know if that looks good.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:23]: And these three are not exactly the same. One is `flutter run`, one is `flutter drive` (and I just noticed a nit, the one in `vmservice_driver.dart` should have an additional `-t` (#61908) and E2E should be both (just updated this)

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: I realized I shouldn't do this because `data` is incremental.

Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: So it seems that I'll have to do the same as the `TimelineSummary` where multiple list generation and sort on the same object is performed... Or I'll have to make a cache system (I don't think I should...). Currently I reverted that moving to constructor change. Let me think about the trade-off.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [reply dev/benchmarks/macrobenchmarks/test/util.dart:212]: I feel sorry too... Here's how I think you might be able to add it back in the future:

Writing that algorithm is a good, what feels unsafe is to make it as an private little function without any unit tests. In the future, if `FrameTimingSummarizer` turns out to be something that we want to add to our API, then we can invest a lot more time in it to improve its ergonomics and efficiency. At that time, we shall certainly make the algorithm as efficient as possible, and add many unit tests to guarantee its correctness. (Right now we're just trying to unblock our critical path asap.)

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:115]: As you pointed out in https://github.com/flutter/flutter/pull/61509#discussion_r457753936, most people who use this function probably don't expect the `List<FrameTiming>` to be mutating in their first thought.

I'd suggest to let your `watcher` record a `List<FrameTiming>`, and give that final list to `FrameTimingSummarizer` after that list is fully recorded.

You can then extract the build and raster duration and sort in the constructor. I feel it's probably still cleaner to call `_findPercentile` in each getter so you don't have to introduce a lot of class member fields.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done with modification to `flutter_test`

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done with refactoring `FrameTimingSummarizer` using factory constructor.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: The internal data of `Duration` is `int` in microseconds.

Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: And it doesn't have a `operator /` method. The only one is `operator ~/ (int)`. Difference is within microseconds which I assume is fine.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:20]: Done with #61958

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: (moved to factory constructor)

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:130]: Nit: when a function has a lot of arguments, Flutter usually put them under `{}` so it would be like the following
```
const FrameTimingSummarizer._({
  this.frameBuildTime,
  ...
});

// Use it with argument name provided to reduce confusion
FrameTimingSummarizer._(
  ...,
  averageFrameBuildTime: frameBuildTime.reduce(add) ~/ data.length,
  ...,
);
```

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (liyuqian) [inline dev/benchmarks/macrobenchmarks/test/util.dart:31]: Nit: as we're moving away from `flutter_driver`, I wonder if `ControlCallback` is a better name than `DriveCallback` as the callback takes a `WidgetController`. CC @dnfield for suggestion too as I imagine that we'll use this in all future e2e performance tests.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (liyuqian) [inline dev/devicelab/lib/tasks/perf_tests.dart:378]: `than` -> `then`

**Review by liyuqian — APPROVED** [review]
LGTM with nits. CC @dnfield to see if there's any extra comments.

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart:130]: `{}` means named arguments, and optional if not marked as `@required`, with default value if not assigned `null`. Here I don't think they should be optional or named since this private constructor is just above its call and should only be used here.

File: dev/devicelab/lib/tasks/perf_tests.dart
Comment (CareF) [reply dev/devicelab/lib/tasks/perf_tests.dart:378]: Done

File: dev/benchmarks/macrobenchmarks/test/util.dart
Comment (CareF) [reply dev/benchmarks/macrobenchmarks/test/util.dart]: Done.

**Pull request merged**
